trio/_socket.py

```diff
@@ -424,18 +521,24 @@ async def _resolve_address_nocp(type, family, proto, *, ipv6_v6only, address, lo
     # The above ignored any flowid and scopeid in the passed-in address,
     # so restore them if present:
     if family == _stdlib_socket.AF_INET6:
-        normed = list(normed)
+        list_normed = list(normed)
         assert len(normed) == 4
+        # typechecking certainly doesn't like this logic, but given just how broad
+        # Address is, it's kind of impossible to write the below without type: ignore
```

Comment:
This might be possible to type, if you define `TypeGuard` functions for the two len checks here, then change each branch to just build a whole tuple instead of converting to list (so we preserve the heterogeneity). Or unpack it into four variables.

Reply:
Ooh, I didn't try with `TypeGuard`. Though I don't think it's terribly important to fix these lines, unless those typeguard functions can be reused elsewhere.